src/openloco/windows/News/News.cpp

```diff
@@ -0,0 +1,875 @@
+        {
+            registers regs;
+            regs.edi = (int32_t)dpi;
+            regs.esi = (int32_t)self;
+            regs.ebp = (int32_t)news;
+            call(0x0042A136, regs);
+        }
+
+        // 0x0042A036
+        static void drawViewportString(gfx::drawpixelinfo_t* dpi, uint16_t x, uint16_t y, uint16_t width, uint8_t itemType, uint16_t itemIndex)
+        {
+            auto args = FormatArguments();
+
+            switch (itemType)
+            {
+                case newsItems::industry:
+                {
+                    auto industry = industrymgr::get(itemIndex);
+                    args.push(industry->name);
+                    args.push(industry->town);
+                    break;
+                }
+
+                case newsItems::station:
+                {
+                    auto station = stationmgr::get(itemIndex);
+                    args.push(station->name);
+                    args.push(station->town);
+                    break;
+                }
+
+                case newsItems::town:
+                {
+                    auto town = townmgr::get(itemIndex);
+                    args.push(town->name);
+                    break;
+                }
+
+                case newsItems::vehicle:
+                {
+                    auto vehicle = thingmgr::get<openloco::vehicle>(itemIndex);
+                    auto company = companymgr::get(vehicle->owner);
+                    if (is_player_company(vehicle->owner))
+                    {
+                        args.push(string_ids::company_vehicle);
+                    }
+                    else
+                    {
+                        args.push(string_ids::competitor_vehicle);
+                    }
+                    args.push(company->name);
+                    args.skip(2);
+                    args.push(vehicle->var_22);
+                    args.push(vehicle->var_44);
+                    break;
+                }
+
+                case newsItems::company:
+                {
+                    auto company = companymgr::get(itemIndex);
+                    args.push(company->name);
+                    break;
+                }
+
+                case 5:
+                case 6:
+                    break;
+
+                case newsItems::vehicleTab:
+                {
+                    auto vehicleObj = objectmgr::get<vehicle_object>(itemIndex);
+                    args.push(vehicleObj->name);
+                    break;
+                }
+            }
+
+            switch (itemType)
+            {
+                case newsItems::industry:
+                case newsItems::station:
+                case newsItems::town:
+                case newsItems::vehicle:
+                case newsItems::company:
+                case newsItems::vehicleTab:
+                {
+                    stringmgr::format_string(byte_112CC04, string_ids::black_tiny_font, &args);
```

Comment:
You have implemented the draw_string_centred_clipped function. We should be keeping that function call.